Youtube player

bernardomjunior opened the pull request:
youtube videos now can be watched without leaving the application

bernardomjunior committed:
Adds youtube api Jar file dependency

Needed to call youtube API

bernardomjunior committed:
Adds youtube view to pich_to_zoom activity's xml and Fixes xml indentation

Youtube's view is located at the same place as PhotoView's one, activity has to hide one or another depending of the situation

bernardomjunior committed:
Updates gitignore of secrets file, where api keys land

bernardomjunior committed:
Adds capability to PinchToZoomActivity to load youtube videos

malkes commented on app/src/main/java/br/com/concrete/tentacle/features/library/loan/LoanViewHolder.kt:
```diff
@@ -42,13 +38,10 @@ LoanViewHolder(item: View) : RecyclerView.ViewHolder(item) {
+                val extras = Bundle()
+                extras.putString(PinchToZoomActivity.ID, video.video_id)
+                extras.putString(PinchToZoomActivity.TYPE, PinchToZoomActivity.VIDEO)
+                startPinchActivity(item.context as AppCompatActivity, extras)
```
É interessante pensar aqui no safe cast. 
Se o contexto não for de uma AppCompatActivity esse cast vai lancar uma exception. Tudo bem que especificamente neste contexto dificilmente isso iria acontecer. Mas não custa previnir

Da  uma sacada https://kotlinlang.org/docs/reference/typecasts.html#safe-nullable-cast-operator

Uma abordagem segura é usar o safe cast e se der sucesso chamar a activity usando a extension como vc fez ou se der nulo usar uma extension de context sem animação por exemplo. Sacou? Só para previnir mesmo

malkes commented on app/src/main/java/br/com/concrete/tentacle/features/library/loan/LoanViewHolder.kt:
```diff
@@ -59,19 +52,26 @@ LoanViewHolder(item: View) : RecyclerView.ViewHolder(item) {
-                        extras = extras,
-                        animation = ActivityAnimation.TRANSLATE_UP
-                    )
+                extras.putString(PinchToZoomActivity.ID, idImage)
```
a criação desse bundle pode ficar um pouco mais Kotlin. Exemplo:

val extras = Bundle().apply {
    putString(PinchToZoomActivity.ID, idImage)
    putString(PinchToZoomActivity.TYPE, PinchToZoomActivity.IMAGE)    
}

malkes commented on app/src/main/java/br/com/concrete/tentacle/features/library/loan/PinchToZoomActivity.kt:
```diff
@@ -1,20 +1,30 @@
+class PinchToZoomActivity : YouTubeBaseActivity() {
 
-class PinchToZoomActivity : AppCompatActivity() {
+    private lateinit var onInitializedListener: YouTubePlayer.OnInitializedListener
```
Se o atributo será usado apenas em um único local pode inicializar ele como val dentro do método mesmo. A não ser seja preciso usar ele em algum outro local além do método initializeVideo

malkes commented on app/src/main/java/br/com/concrete/tentacle/features/library/loan/PinchToZoomActivity.kt:
```diff
@@ -51,4 +82,27 @@ class PinchToZoomActivity : AppCompatActivity() {
+            override fun onInitializationSuccess(
+                provider: YouTubePlayer.Provider?,
+                youtubePlayer: YouTubePlayer?,
+                p2: Boolean
```
É uma boa prática dar nomes mais significativos para as variáveis. Neste caso aqui imagino que o próprio Android Studio criou assim mas ai podemos ir na documentação e ver o nome ou o que o método faz e dar um nome mais significativo. 

https://developers.google.com/youtube/android/player/reference/com/google/android/youtube/player/YouTubePlayer.OnInitializedListener.html?hl=pt-BR

no caso do p2 o nome dele é wasRestored. No link explica o que ele indica.

malkes commented on app/src/main/java/br/com/concrete/tentacle/features/library/loan/PinchToZoomActivity.kt:
```diff
@@ -43,6 +49,31 @@ class PinchToZoomActivity : AppCompatActivity() {
+        val imageId = intent.getStringExtra(ID)
+        imageId?.let {
+            photoView.loadImageUrl(
+                Utils.assembleGameImageUrl(
```
Se vc você refatorar, como você acha que seria uma forma mais elegante de fazer esse assembleGameImageUrl ?

bernardomjunior replied:
Prefiro extension em relação à utils. é essa a pergunta?

malkes commented on app/src/main/java/br/com/concrete/tentacle/features/library/loan/PinchToZoomActivity.kt:
```diff
@@ -43,6 +49,31 @@ class PinchToZoomActivity : AppCompatActivity() {
+    }
+
+    private fun loadVideo() {
+        photoView.visible(false)
```
Aqui é uma opinião pessoal.

Hoje eu prefiro criar extensions para cada ação. Exemplo:

visible()
invisible()
gone()

e uma que aceita um booleando para casos onde da para usar. 
visibleOrGone(result)
visibleOrInvisible(result)

Por que isso?
Porque fica fácil de ler o código e entender o que está acontecendo e a chance de errar é menor também. Sacou?

Mas é uma questão bem pessoal

bernardomjunior replied:
O que tu achas de `view.visibility = View.Gone`?

malkes replied:
Então. É muito pessoal isso. Particularmente eu prefiro as coisas mais diretas, onde qualquer pessoa bate o olho e já sabe o que está acontecendo e não tem que pensar.
Seu eu olhar para um `button.visible()` ou para o `button.gone()` eu já sei o que vai acontecer com eles. Seu olho para um `button.visibleOrGone(result)` eu já sei o comportamento dele vai depender o result. Saca?

malkes commented on .gitignore:
```diff
@@ -24,3 +24,4 @@ Thumbs.db
 build/
 app/google-services.json
 /.project
+/app/src/main/res/values/secrets.xml
```
Depois me manda a chave para eu testar aqui tb

bernardomjunior replied:
belê

malkes commented on app/src/main/res/layout/activity_pinch_to_zoom.xml:
```diff
@@ -1,33 +1,46 @@
-            app:layout_constraintRight_toRightOf="parent"
-            app:layout_constraintTop_toTopOf="parent"/>
+        android:id="@+id/btClose"
+        android:layout_width="@dimen/dp_25"
```
Hoje eu não usaria esse tipo de dimen. Tem dois pontos aqui.
1 - Não é recomendado o uso de tamanhos impares
2 - Não tem diferença colocar 25dp hardcoded ou @dimen/dp_25

A melhor abordagem nesse caso seria a criação de padrões. Exemplo:

default = 8dp
medium = 16dp
large = 32dp

os nomes são apenas sugestões, vai depender do alinhamento dom designer e dev team.
A vantagem disso é que se no futuro o meu default não for mais 8dp eu simplesmente vou mudar o valor dele no dimens e refletir em todo o app. Agora imagina se o preciso dizer que onde tem dp_25 agora deve ter do_30. Ficaria muito ruim mudar o valor do dp_35 para 30 né. Ou muito trabalhoso ter que ficar  trocando em todo o código. Sacou??

bernardomjunior replied:
saquei, então devo rever o código feito antes e fazer alterações que o melhore, certo?

malkes replied:
Não, por enquanto não. Podemos pensar mais para a frente em refatorar tudo para um módulo de design por exemplo. Mas por hora usa assim mesmo. Só comentei para ficar no seu radar.

malkes reviewed (commented):
No geral ficou tudo muito bom. O Kotlin está bem dentro do Code Style mesmo.
Fiz alguns comentários pontuais e comentei algumas coisas que já estavam no projeto mas que hoje eu já faria bem diferente. Da uma olhada e qualquer coisa a gente bate um papo depois para trocar uma ideia e refletir sobre os pontos.

bernardomjunior committed:
Guarantees null safety launching Pinch Activity from Loan View Holder

bernardomjunior committed:
Changes bundle creation to Kotlin-ish

bernardomjunior committed:
Moves youtube's OnInitializedListener to local variable

bernardomjunior committed:
Fixes variable naming

bernardomjunior committed:
Removes unneeded util method

Method used to formulate a youtube webpage link, now it is not needed because of video is now being processed by api
